### PR TITLE
Add extra reporter hooks to track requirements, pins and backtracks

### PR DESCRIPTION
**Pull request by pfmoore**

### CODE REVIEW

File: src/resolvelib/resolvers.py [hunk 1]
```diff
@@ -224,6 +226,7 @@ def _backtrack(self):
 
             # Retract the last candidate pin, and create a new (b).
             name, candidate = self._states.pop().mapping.popitem()
+            self._r.backtrack(candidate)
```
Comment (uranusjr) [inline src/resolvelib/resolvers.py:229]: Would `backtracking` be more consistent?

File: src/resolvelib/reporters.py [hunk 2]
```diff
@@ -22,3 +22,15 @@ def ending_round(self, index, state):
     def ending(self, state):
         """Called before the resolution ends successfully.
         """
+
+    def adding_requirement(self, requirement):
+        """Adding a new requirement into the resolve criteria.
+        """
```
Comment (pradyunsg) [inline src/resolvelib/reporters.py:28]: nit: Documentation string phasing consistency would be nice.

File: src/resolvelib/resolvers.py (hunk 1 above)
Comment (pradyunsg) [reply src/resolvelib/resolvers.py:229]: It would be correct; given that the method is called `backtracking`. ;P

Comment (pfmoore) [reply src/resolvelib/resolvers.py:229]: Point taken :-)

File: src/resolvelib/reporters.py (hunk 2 above)
Comment (pfmoore) [reply src/resolvelib/reporters.py:28]: Not a nit, it's an important point and one I've just not go round to. Will fix this.

File: src/resolvelib/resolvers.py (hunk 1 above)
Comment (pfmoore) [reply src/resolvelib/resolvers.py:229]: Sigh. Simple examples don't backtrack. Thanks for spotting this :-)

Comment (uranusjr) [reply src/resolvelib/resolvers.py]: I’ve been thinking about it for a while, this could be the tipping point I start adding type hints to the code base.

**Pull request merged**